### PR TITLE
Offline Distillation via DistillKit (Part One - Compression Helpers for Logit Capture)

### PR DESCRIPTION
**Pull request by wolfecameron**
## Summary

This PR is **Part 1** of a large effort to add support for offline distillation into OpenInstruct

The full distillation rollout will add:
- teacher logit sampling with vLLM.
- student offline distillation training.
- distillation data handling/loss infrastructure and launch scripts.

This Part 1 PR intentionally scopes to the **compression primitives foundation only** (no sampling/training pipeline yet).

## PR Scope

Adds the core compression module needed by later distillation stages:

- `open_instruct/distillkit/compression/config.py` for distribution quantization config/types and validation.
- `open_instruct/distillkit/compression/bitpack.py` for bit-packing and unpacking token index payloads.
- `open_instruct/distillkit/compression/monotonic_logprobs.py` for monotonic logprob compression/decompression utilities.
- `open_instruct/distillkit/compression/compressor.py` for sparse top-k compression/decompression wrapper logic.
- `open_instruct/distillkit/compression/__init__.py` and `open_instruct/distillkit/__init__.py` minimal exports for this scoped subset.
- `configs/compression/default.yml` default compression config to be consumed by upcoming sampling/training stages.
- `open_instruct/test_distillkit_compression.py` focused unit tests covering bitpack roundtrip, config roundtrip, and exact-only sparse compressor roundtrip.

## Attribution
- Contribution work from Netflix Research: Cameron Wolfe (`cameronwolfe@netflix.com`)
- DistillKit adaptation: Portions of the compression implementation were adapted from [DistillKit](https://github.com/arcee-ai/DistillKit), with attribution headers preserved in the relevant ported files.

### CODE REVIEW

**Comment by finbarrtimbers** [comment]
Thanks for splitting the original PR! Taking a look.

File: open_instruct/distillkit/compression/bitpack.py
Comment (finbarrtimbers) [inline open_instruct/distillkit/compression/bitpack.py:1]: Why can't we have distillkit as a library and import this stuff?

File: open_instruct/distillkit/compression/bitpack.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
+# Portions of this file were adapted from distillkit (github.com/arcee-ai/DistillKit)
+# Copyright 2025 Arcee AI. Licensed under the Apache License, Version 2.0.
+
+import torch
+
+
+def pack_to_bytes(x: torch.LongTensor, elem_bits: int) -> torch.ByteTensor:
+    """Pack integer tensor values into byte tensors."""
+    assert 1 <= elem_bits <= 64, "elem_bits must be between 1 and 64"
```
Comment (finbarrtimbers) [inline open_instruct/distillkit/compression/bitpack.py]: Please use a `ValueError` for argument validation!

File: open_instruct/distillkit/compression/bitpack.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+    mask = (1 << elem_bits) - 1
```
Comment (finbarrtimbers) [inline open_instruct/distillkit/compression/bitpack.py]: Inline this, as it's only used once, and provide a comment describing what it's doing?

File: open_instruct/distillkit/compression/bitpack.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    x = x & mask
+
+    bit_positions = torch.arange(elem_bits - 1, -1, -1, device=x.device)
+    bits = ((x.unsqueeze(-1) >> bit_positions) & 1).to(torch.uint8)
+
+    original_shape = x.shape
```
Comment (finbarrtimbers) [inline open_instruct/distillkit/compression/bitpack.py]: Inline, as it's only used once!

File: open_instruct/distillkit/compression/bitpack.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    bits = bits.view(*original_shape[:-1], -1)
+
+    total_bits = bits.size(-1)
```
Comment (finbarrtimbers) [inline open_instruct/distillkit/compression/bitpack.py]: Inline as it's only used once!

File: open_instruct/distillkit/compression/bitpack.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    pad_length = (8 - (total_bits % 8)) % 8
+    if pad_length > 0:
+        bits = torch.nn.functional.pad(bits, (0, pad_length))
+    bits = bits.contiguous()
```
Comment (finbarrtimbers) [inline open_instruct/distillkit/compression/bitpack.py]: Combine these two lines?

File: open_instruct/distillkit/compression/bitpack.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    bit_positions = torch.arange(elem_bits - 1, -1, -1, device=x.device)
+    bits = ((x.unsqueeze(-1) >> bit_positions) & 1).to(torch.uint8)
+
+    original_shape = x.shape
+    bits = bits.view(*original_shape[:-1], -1)
+
+    total_bits = bits.size(-1)
+    pad_length = (8 - (total_bits % 8)) % 8
+    if pad_length > 0:
```
Comment (finbarrtimbers) [inline open_instruct/distillkit/compression/bitpack.py]: I think that we don't need the check, as if `pad_length` is 0, it's a no-op?

File: open_instruct/distillkit/compression/bitpack.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    bits = bits.contiguous()
+
+    bits = bits.view(*original_shape[:-1], -1, 8)
+    power = torch.tensor([128, 64, 32, 16, 8, 4, 2, 1], dtype=torch.uint8, device=x.device)
+    return (bits * power).sum(dim=-1).to(torch.uint8)
+
+
+def unpack_from_bytes(bytes_tensor: torch.ByteTensor, elem_bits: int, original_num_elements: int) -> torch.LongTensor:
+    """Unpack byte tensors into integer tensor values."""
+    assert 1 <= elem_bits <= 64, "elem_bits must be between 1 and 64"
```
Comment (finbarrtimbers) [inline open_instruct/distillkit/compression/bitpack.py]: ValueError please!

File: open_instruct/distillkit/compression/bitpack.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        raise ValueError(
+            f"Need {total_bits_needed} bits for {original_num_elements} elems, only have {total_bits_available}."
+        )
+
+    bit_positions = torch.arange(7, -1, -1, device=bytes_tensor.device)
+    bits = ((bytes_tensor.unsqueeze(-1) >> bit_positions) & 1).to(torch.uint8)
+    bits_flat = bits.view(*original_shape[:-1], -1)
+    bits_needed = bits_flat[..., :total_bits_needed]
+
+    new_shape = list(original_shape[:-1]) + [original_num_elements, elem_bits]
```
Comment (finbarrtimbers) [inline open_instruct/distillkit/compression/bitpack.py]: Is the `list(...)` needed? Can you try removing it?

File: open_instruct/test_distillkit_compression.py
Comment (finbarrtimbers) [inline open_instruct/test_distillkit_compression.py:1]: 1. Can you move this to `distillkit/compression/test_compression.py`? 
2. Can you use `unittest` instead of `pytest`? (That's what we use for the rest of our code.)  We also use the `parameterize` library for parameterization.

File: open_instruct/distillkit/compression/compressor.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+from open_instruct.distillkit.compression.bitpack import pack_to_bytes, unpack_from_bytes
+from open_instruct.distillkit.compression.config import DistributionQuantizationConfig
+from open_instruct.distillkit.compression.monotonic_logprobs import (
+    compress_monotonic_logprobs,
+    decompress_monotonic_logprobs,
+)
+
+
+class LogprobCompressor:
```
Comment (finbarrtimbers) [inline open_instruct/distillkit/compression/compressor.py:14]: Should there be a base `Compressor` class that this is implementing that it could subclass from?

File: open_instruct/distillkit/compression/config.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+
+import torch
+
+
+class SpecialTerm(Enum):
+    SQRT = "sqrt"
+    EXP = "exp"
+
+
+class TermDtype(Enum):
```
Comment (finbarrtimbers) [inline open_instruct/distillkit/compression/config.py]: Why not use the torch dtypes directly?

File: open_instruct/distillkit/compression/config.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    FLOAT16 = "float16"
+    BFLOAT16 = "bfloat16"
+    FLOAT32 = "float32"
+    FLOAT64 = "float64"
+
+    def bit_width(self) -> int:
+        return 16 if self in {TermDtype.FLOAT16, TermDtype.BFLOAT16} else (32 if self == TermDtype.FLOAT32 else 64)
```
Comment (finbarrtimbers) [inline open_instruct/distillkit/compression/config.py]: Nit: Could be 

```
def bit_width(self) -> int:
    return int(self.value.split("float")[-1])
```

(and with a `assert float in self.value, f'We only support floats, not {self.value}'` if desired.) 

If we switch to using the torch dtypes directly, could just be `return 8 * value.itemsize()`.

File: open_instruct/distillkit/compression/config.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+
+
+@dataclass
+class QuantizationBin:
+    scale_dtype: TermDtype
+    element_bits: int
+    num_elements: int
+
+    @classmethod
+    def from_dict(cls, data: dict) -> "QuantizationBin":
```
Comment (finbarrtimbers) [inline open_instruct/distillkit/compression/config.py]: Is this needed? I think you [can already do](https://colab.research.google.com/drive/1yegf3GnX_eVwcQu_oN2RTDomYWGjeSrG?usp=sharing):

```
QuantizationBin(**{'scale_dtype': torch.bfloat16, 'element_bits': 10, 'num_elements': 3})
```

File: open_instruct/distillkit/compression/config.py [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+    residual_bins: list[QuantizationBin] = field(default_factory=list)
+    delta_encoding: bool = True
+    error_diffusion: bool = False
+    normalize_t: bool = False
+
+    @classmethod
+    def from_dict(cls, data: dict) -> "DistributionQuantizationConfig":
+        terms_raw = data.get("polynomial_terms")
+        terms: list[SpecialTerm | int] | None
+        if terms_raw is None:
```
Comment (finbarrtimbers) [inline open_instruct/distillkit/compression/config.py:81]: This is a bit dense, can you add a comment here?

File: open_instruct/distillkit/compression/config.py [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+        if self.k < self.exact_k:
+            raise ValueError("k must be >= exact_k")
+        if self.exact_k < self.k and (not self.polynomial_terms) and (not self.residual_bins):
+            raise ValueError("If exact_k < k, polynomial_terms or residual_bins must be provided.")
+        approx_terms = self.k - self.exact_k
+        bin_elems = sum(bin_.num_elements for bin_ in self.residual_bins)
+        if bin_elems > approx_terms:
+            raise ValueError("residual_bins total num_elements must be <= k - exact_k")
+
+    def to_dict(self) -> dict:
```
Comment (finbarrtimbers) [inline open_instruct/distillkit/compression/config.py:131]: Can we just use `dataclasses.asdict`?

File: open_instruct/distillkit/compression/monotonic_logprobs.py
Comment (finbarrtimbers) [inline open_instruct/distillkit/compression/monotonic_logprobs.py:1]: Please add docstrings here!

File: open_instruct/distillkit/compression/bitpack.py
Comment (wolfecameron) [reply open_instruct/distillkit/compression/bitpack.py:1]: Distillkit isn't a library, so you can't pip install it. If that would be easier, I can reach out to the authors and try to publish it as a package.

File: open_instruct/distillkit/compression/bitpack.py (hunk 6 above)
Comment (wolfecameron) [reply open_instruct/distillkit/compression/bitpack.py]: you're right, fixed and added a test case for this

File: open_instruct/distillkit/compression/compressor.py (hunk 9 above)
Comment (wolfecameron) [reply open_instruct/distillkit/compression/compressor.py:14]: Definitely can if you'd prefer - distillkit only has logprob compressor, so there is currently no other classes that would inherit this. This could be helpful for extending in the future.

File: open_instruct/distillkit/compression/config.py [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+        return 16 if self in {TermDtype.FLOAT16, TermDtype.BFLOAT16} else (32 if self == TermDtype.FLOAT32 else 64)
+
+    def dtype(self) -> torch.dtype:
+        mapping = {
+            TermDtype.FLOAT16: torch.float16,
+            TermDtype.BFLOAT16: torch.bfloat16,
+            TermDtype.FLOAT32: torch.float32,
+            TermDtype.FLOAT64: torch.float64,
+        }
+        return mapping[self]
```
[the remark this answers is not in the document]
Comment (wolfecameron) [reply open_instruct/distillkit/compression/config.py]: stale after making changes below

File: open_instruct/distillkit/compression/config.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (wolfecameron) [reply open_instruct/distillkit/compression/config.py]: stale after making changes below

File: open_instruct/distillkit/compression/bitpack.py [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+    bits_needed = bits_needed.contiguous().view(*new_shape)
+
+    powers = 2 ** torch.arange(elem_bits - 1, -1, -1, device=bits_needed.device)
```
[the remark this answers is not in the document]
Comment (wolfecameron) [reply open_instruct/distillkit/compression/bitpack.py:51]: probably doesn't make sense because this vector is pretty small and not worth caching

File: open_instruct/distillkit/compression/bitpack.py [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+    bits = bits.view(*original_shape[:-1], -1)
+
+    total_bits = bits.size(-1)
+    pad_length = (8 - (total_bits % 8)) % 8
+    if pad_length > 0:
+        bits = torch.nn.functional.pad(bits, (0, pad_length))
+    bits = bits.contiguous()
+
+    bits = bits.view(*original_shape[:-1], -1, 8)
+    power = torch.tensor([128, 64, 32, 16, 8, 4, 2, 1], dtype=torch.uint8, device=x.device)
```
[the remark this answers is not in the document]
Comment (wolfecameron) [reply open_instruct/distillkit/compression/bitpack.py:25]: tiny tensor so probably doesn't matter much

File: open_instruct/distillkit/compression/bitpack.py
Comment (finbarrtimbers) [reply open_instruct/distillkit/compression/bitpack.py:1]: Ah that's silly. I would appreciate that if you have time! Let's leave that as a TODO though so as to not block this PR?

File: open_instruct/distillkit/compression/compressor.py (hunk 9 above)
Comment (finbarrtimbers) [reply open_instruct/distillkit/compression/compressor.py:14]: Ah nah that's fine then.

File: open_instruct/distillkit/compression/bitpack.py
Comment (wolfecameron) [reply open_instruct/distillkit/compression/bitpack.py:1]: Sure! Let me ping them, I can work on this in the background. I agree this is the best approach because the changes we made here would also be good to make in the original distillkit.
